### ISSUE TITLE
如果我想在工作中运用机器学习，我应该怎么学习？

### ISSUE THREAD

**Issue by DerekGrant** [DerekGrant/Union-A---XLab#1]
本文内容收集自优达学城机器学习/深度学习互助班助教团讨论记录和我与朋友的讨论

很多优达学员学习机器学习往往是为了找一份机器学习的工作或者想让自己能在工作中用上机器学习。所以往往会有疑问，我们学完这个课程后能找到一份工作吗？如果我想在工作中使用机器学习，我需要学到什么程度？有学员只想学习一下“实用技能”，例如Sklearn和TensorFlow的使用，只想学习深度学习，而对各种其他算法及其推导不感兴趣。而在机器学习领域，各种公式理论充斥在教材其中，似乎无法忽略。那我们如果只是想找一份机器学习的工作，而不是想从事这方面的科研，那我还要学习机器学习的理论吗，要学多深呢？我们在平时的学习中如何有选择的学习呢？我们的助教团针对这个问题展开了讨论。

# 话题一：如果我只是想应用机器学习，而非科研，我们需要学习理论知识吗？理论知识有什么用吗？

有的学员认为自己工作中使用的算法并不需要复杂的理论，往往调用Sklearn的各种算法比较一下结果就好，使用各种理论知识调参收效甚微，认为理论知识学习太深似乎没太多作用，自己不从事相关科研，不需要浪费时间。

对于这个问题，**机器学习互助班15期助教，深度学习工程师高民权**是这样认为的：

我认为知道这些原理还是有必要的，不需要像一个nerd一样把条条框框都知道，比方说咣咣咣把SVM几种kernel的优劣背出来，这种，但是其中的原理还是很有借鉴价值的。例如，KNN利用周围来决策某个点，在很多时候，我们需要处理好几个相互作用的东西，这个在文本或者时间序列中很常见，我们为了更好的衡量某个点，就可以采用“类似于”knn的思想，但是又不是纯的Knn，那么这个时候，如果一个人对knn原理不清楚，那么可能就想不到这个方法，另外，就算了解原理，但是实现方法完全没有思考过，那么就算如此简单的一个程序，要编码实现也是有难度的，也会漏洞百出，所以了解原理的第一个原因就是，为了处理新的，没有遇到过的问题。有句话说的好，“靡革匪因，糜故匪新”，没有什么革新不是基于以前的东西，也没有什么以前的，固有的东西不包含新的思想；第二个原因就是，尤其是当我们在做深度学习的时候，变量非常多，从数据的分部，到网络结构，初始化，参数更新的方式，需要自己定义的太多，这个时候，如果不熟悉原理，那么遇到问题，例如loss震荡太严重，loss不下降，loss看起来下降，但实际上快速下降到某个点之后就不下降，运行太慢，overfitting太大，等等，要是不知道原理，那就会真的像无头苍蝇一样，各种尝试，而不能analytic，而且，很多问题，尤其公司的研究岗位，遇到的问题也常常是没有现场解决方案的，所以就更需要出现问题的时候能自己找到思路了，要是不理解原理，就不知道顺着那条路去走。虽然说，形之上者谓之道，形之下者谓之器，但是我认为对于各行各业，尤其对于我们搞计算机的人，这两者是紧密不可分割的，只有这样，才能站在别人的肩膀上，去把心思花在解决新问题上。

我的朋友，**某知名团队深度学习工程师**：

推导SVM之类的我觉得还是有必要的经历，我现在虽然也不一定记得很清楚，可是我在上NG课程的时候，就知道了其实是去求一个最小间隔的，而这种类型的损失函数其实应用到了深度学习当中，其实就是hinge loss，所以我是认为这些更多的是在培养你的一种research sense，可能其意义已经超过了知识点本身。至于后面的问题嘛，就是道行深不深了，就像我说的import tensorflow as tf的人一大把，又有多少真的理解了backprop呢。我始终认为data clean是非常重要的，并且传统的机器学习一定要涉及而不能够只会深度学习。

## 机器学习互助班18期助教，工业机器人工程师耿庆祥：

其实我想举个例子，我是自动化专业的，我们专业知识其实都快忘光了，但是我还有大部分的印象，看到任何一个设备，一个事物，甚至是软件，生活，情感，都会用控制反馈的主题思想去理解它，它是开环闭环的，它能反馈什么，整个系统大概如何工作，感觉自动化的事物大框架特别的清晰，看到新东西不会有懵逼的感受。其他专业的可能也会有类似的影响，这个我不太确定。但是机器学习，能让我们对数据有个新的认识。我们曾经都是靠物理世界的模型去理解世界，突然有一种新的东西，从数据的角度去理解世界，这是一个新的窗户，我觉得这种新的思路其实是最重要的。ALL model are wrong，but some of them are useful.物理模型跟数据推断出来的结果有可能还是非常有用的。相当于有个新的工具去认识事物。机器学习理论的指导作用恰就在这里，它帮你理解整个机器学习算法的运行，带给你一种潜移默化的思考方式。

## 机器学习互助班17期助教，云脑工程师汤奎桦：

在实际工程应用中，简单地调包是不能解决问题的。因为很多时候工程问题的环境还是比较苛刻的，需要你对算法做很多的实现和优化，而这就需要你对算法的透彻理解。比如项目会需要你用特定的语言（如C++）实现一个算法，并进行运行效率等方面的优化，而这些符合特定需求的算法是没有现成的实现的，需要你自己去完成。

机器学习互助班25期助教，数据挖掘工程师刘雨涛：

在很多工程应用中，因为版权的原因，我们是不能直接调包的，必须亲自实现。这就需要我们完全理解算法的每一个步骤环节。

## 机器学习互助班6期助教，高校讲师孙斌：

在我班上学员实际的工作面试中，我发现公司并不止需要学员对算法有一个普通的认识和理解，还会考察学员对算法的推导。例如PCA背后的算法，SVM的推导等。

# 话题二：如果以工作为目标，我们对于理论知识应该学多深？

## 耿庆祥：

对于理论应该掌握到什么程度，我想尽力深挖应该永远不会错。平时学习照顾到面的问题，具体工作深入到点的问题。在学习时候以面为主，保证有深入点的能力。SVM要不要看公式，要不要推到，必须的，要不要理解每一个参数？要！但是说半个月以后要求你公式倒被如流，我觉得有点高，但是半个月以后就讲不出来核函数是干嘛用的，是升维了，还是降维了，这就白学了。

## 孙斌：

总体来说，理论学得越深越好，但是不同的学员可以在和自己相关领域深挖。

## 机器学习互助班3期助教，算法工程师侯宇

原则上学得越深越好，但是希望自己不光懂理论，还要能够自己将算法实现一遍。如果不知道自己需要学到什么层次的，可以先明确自己的目标，和互助班助教沟通确定。如果有找工作需求的学员，可以先去公司面试下试试，看自己离目标的公司要求差距多少，集中补充那些欠缺的部分。我们在学习的过程中，学到的算法一定要去动手实现下。这个过程中你会发现书本上的算法在真正使用的时候会有很大的差别，需要考虑很多更复杂的问题，理解得更深入。比如说我们平时学习梯度下降，但在实际运用的时候用的是随机梯度下降SGD，这个变化我们就是需要深入理解的。在学某个章节的时候，我们理解一些必要的点就足够了，其他根据个人需求再深入。比如SVM，我们知道什么是硬间隔、软间隔、核函数，然后知道SVM最大的问题就是计算慢，理解SOM如何加速计算就足够了，这就是精髓部分。

## 机器学习互助班11期助教，航天工程师石磊

我觉得就很多基础比较薄弱的学员来说，可以分应用层和理论层来学习。理论层知道一些基础的算法已经算法间的关系区别，建立一个整体的框架和知识体系，应用层了解Sklearn等工具的使用，有更多时间的话再在之前的基础上深化对理论和代码的理解。很多学员会在一个算法上停留过久，却没了解其他方法，容易只见树叶不见森林。

总的来说，理论知识还是很重要的。如果不深入掌握理论，你只能在一些固定的场景使用机器学习，而无法在复杂多变的现实环境中灵活应对，修改优化算法达到工程需要。而在学习过程中，我们首先要明确自己的学习目标。可以和助教交流，也可以去面试找找差距。在有限的学习过程中，我们尽量优先了解足够多的算法，搭建起一个知识框架，然后再针对和自己需求相关的部分深挖。我们要同等重视理论和代码实现的重要性，理论的深入理解有助于工程实践，而算法实现能帮助自己通盘掌握算法细节。即使我们只是想学深度学习，也有必要学习传统机器学习，这是一整个知识体系，我们只有全面了解，才能应对自如。

**Comment by idisblueflash** [comment]
建议增加样式，看得清楚。 像这样：

### 机器学习互助班11期助教，航天工程师石磊

我觉得就很多基础比较薄弱的学员来说，可以分应用层和理论层来学习。理论层知道一些基础的算法已经算法间的关系区别，建立一个整体的框架和知识体系，应用层了解Sklearn等工具的使用，有更多时间的话再在之前的基础上深化对理论和代码的理解。很多学员会在一个算法上停留过久，却没了解其他方法，容易只见树叶不见森林。

**Comment by DerekGrant** [comment]
@idisblueflash 谢谢，已修改！
